docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+staking contract can record their operator, beneficiary, and authorizer. Since
+these values are read-only, they should be copied from KEEP staking contract to
+T staking contract and NU stakers should set them during the delegation. For
+existing stakes, the entire stake amount is registered in T staking contract.
+
+There are 2 hours to cancel a stake without suffering an undelegation period.
+This functionality is needed e.g. in case of a typo in the operator address.
+Undelegation can not be canceled in the same transaction in which it was
+created. Once the delegation has been canceled, the operator address can be used
+again.
+
+==== Authorizing an application
+
+Before the operator is eligible to participate in the given application, the
+authorizer appointed during the stake delegation needs to review the application
+and approve it to use the stake. From the moment of approval, the application
+is authorized to slash or seize the stake.
+
+The authorizer decides up to which amount the given application is eligible to
+slash. This allows managing the risk profile of the staker individually for each
+application. The authorized stake amount can only be increased.
+
+Authorization can not be canceled.
+
+The application being authorized is notified via a callback every single time
+new authorization happens and the application may decide to revert the
+transaction if individual requirements of the application are not met. Such
+requirements could be, for example, not enough stake authorized for the
+particular application or not supported legacy stake (see Limitations section).
+
+All applications first need to be authorized in a registry of T applications
+controlled by the governance.
+
+==== Stake top-up
+
+Top-up increases the amount of tokens locked in a delegation. This increases the
+probability of being chosen for work in the network but is only effective for
+future checks of the stake amount.
+
+Top-ups can be executed for native T stakes and for legacy KEEP/NU stakes.
+
+Native T stakers can only top-up their stakes with a liquid T.
+
+Existing KEEP and NU stakers wanting to execute a top-up have two options. One
+option is to wrap their KEEP/NU to T and then, execute a top-up in T staking
+contract. The second option is to execute a top-up in their legacy staking
+contracts and notify T staking contract about the fact their legacy stake
+increased.
+
+Effectively, it means that existing KEEP stakers can mix their legacy KEEP
+stakes with liquid T stakes. Similarly, existing NU stakers can mix their legacy
+NU stakes with liquid T stakes. This functionality adds some complexity to the
+staking contract but it puts existing KEEP/NU stakers in the same spot as new T
+stakers. Without it, existing stakers would not be able to top-up their stakes
+with T earned from operating in the network, so they would be in a worse spot
+than new T stakers allowed to top-up their stakes using earned T and this way
+increasing their rewards.
```

Comment:
If we permit stake-mixing we need to consider what to do when slashing, or when the legacy stake is undelegated but the T stake is not. Permitting the latter case (if the topped-up T stake exceeds the minimum) would let operators move their stakes from legacy to T gracefully without excessive interruption.

Reply:
[From here](https://github.com/threshold-network/solidity-contracts/pull/15#discussion_r704315662)

> Slashing T first helps avoiding legacy contract calls but slashing KEEP/NU first might easen the transition to T over time.

@pdyraga are we expecting honest operators getting slashed to be common enough that we should cater to this specifically?

Reply:
Nah, I do not expect any slashes for honest operators. That's actually a good question, I do not think the situation I was worried about is realistic. 

So I think:
- slash T stake first (for simplicity)
- figure out the way to undelegate from the legacy contract without undelegating from the new contract and not being slashed for stake discrepancy 

could be the way to go.

Reply:
If we permit authorisation decreases, we could require that the operator has reduced their authorisations on all apps to or below the T stake they have before being allowed to initiate undelegation on KEEP. Undelegating without reducing the authorisations first would be treated as a discrepancy.

Reply:
Good idea!